### PR TITLE
[clang] constexpr built-in elementwise add_sat/sub_sat functions. 

### PR DESCRIPTION
**Pull request by c8ef**
Part of #51787.

This patch adds constexpr support for the built-in elementwise add_sat and sub_say functions.

### CODE REVIEW

File: clang/docs/ReleaseNotes.rst [hunk 1]
```diff
@@ -414,6 +414,7 @@ Non-comprehensive list of changes in this release
 - ``__builtin_reduce_or`` and ``__builtin_reduce_xor`` functions can now be used in constant expressions.
 - ``__builtin_elementwise_popcount`` function can now be used in constant expressions.
 - ``__builtin_elementwise_bitreverse`` function can now be used in constant expressions.
+- ``__builtin_elementwise_add_sat`` function can now be used in constant expressions.
```
Comment (cor3ntin) [inline clang/docs/ReleaseNotes.rst:417]: Can we merge all of these?
"The following builtins can now be used in constant expressions < nested bulleted list>"

File: clang/lib/AST/ExprConstant.cpp [hunk 2]
```diff
@@ -11339,6 +11339,31 @@ bool VectorExprEvaluator::VisitCallExpr(const CallExpr *E) {
...
+    QualType DestEltTy = E->getType()->castAs<VectorType>()->getElementType();
+    unsigned SourceLen = SourceLHS.getVectorLength();
+    SmallVector<APValue, 4> ResultElements;
+    ResultElements.reserve(SourceLen);
+
+    for (unsigned EltNum = 0; EltNum < SourceLen; ++EltNum) {
+      APSInt LHS = SourceLHS.getVectorElt(EltNum).getInt();
+      APSInt RHS = SourceRHS.getVectorElt(EltNum).getInt();
+      switch (E->getBuiltinCallee()) {
+      case Builtin::BI__builtin_elementwise_add_sat:
```
Comment (cor3ntin) [inline clang/lib/AST/ExprConstant.cpp:11357]: Can that be anything else here? Shouldn't we just assert or remove the test entirely?

Comment (RKSimon) [reply clang/lib/AST/ExprConstant.cpp:11357]: It'd be great if you could handle subsat in this patch as well. Same when you come to max/min - better to handle the instruction pairs together.

File: clang/test/Sema/constant_builtins_vector.cpp [hunk 3]
```diff
@@ -822,3 +822,19 @@ static_assert(__builtin_elementwise_bitreverse(0x12345678) == 0x1E6A2C48);
 static_assert(__builtin_elementwise_bitreverse(0x0123456789ABCDEFULL) == 0xF7B3D591E6A2C480);
 static_assert(__builtin_bit_cast(unsigned, __builtin_elementwise_bitreverse((vector4char){1, 2, 4, 8})) == (LITTLE_END ? 0x10204080 : 0x80402010));
 static_assert(__builtin_bit_cast(unsigned long long, __builtin_elementwise_bitreverse((vector4short){1, 2, 4, 8})) == (LITTLE_END ? 0x1000200040008000 : 0x8000400020001000));
+
+static_assert(__builtin_elementwise_add_sat(1, 2) == 3);
+static_assert(__builtin_elementwise_add_sat(1U, 2U) == 3U);
+static_assert(__builtin_elementwise_add_sat(~(1 << 31), 42) == ~(1 << 31));
+static_assert(__builtin_elementwise_add_sat((1 << 31), -42) == (1 << 31));
+static_assert(__builtin_elementwise_add_sat(~0U, 1U) == ~0U);
+static_assert(__builtin_bit_cast(unsigned, __builtin_elementwise_add_sat((vector4char){1, 2, 3, 4}, (vector4char){4, 3, 2, 1})) == (LITTLE_END ? 0x05050505 : 0x05050505));
```
Comment (RKSimon) [inline clang/test/Sema/constant_builtins_vector.cpp:831]: Avoid uniform vector results to better show each element is working.

**Review by cor3ntin — APPROVED** [review]
I am happy if @RKSimon is

**Review by RKSimon — APPROVED** [review]
LGTM cheers

**Pull request merged**